hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,348 @@
+//! [ascent] datalog implementation of analysis.
+
+use ascent::lattice::BoundedLattice;
+use itertools::Itertools;
+
+use hugr_core::extension::prelude::{MakeTuple, UnpackTuple};
+use hugr_core::ops::{OpTrait, OpType};
+use hugr_core::{HugrView, IncomingPort, Node, OutgoingPort, PortIndex as _, Wire};
+
+use super::value_row::ValueRow;
+use super::{AbstractValue, AnalysisResults, DFContext, PartialValue};
+
+type PV<V> = PartialValue<V>;
+
+/// Basic structure for performing an analysis. Usage:
+/// 1. Get a new instance via [Self::default()]
+/// 2. (Optionally / for tests) zero or more [Self::prepopulate_wire] with initial values
+/// 3. Call [Self::run] to produce [AnalysisResults]
+pub struct Machine<V: AbstractValue>(Vec<(Node, IncomingPort, PartialValue<V>)>);
+
+/// derived-Default requires the context to be Defaultable, which is unnecessary
+impl<V: AbstractValue> Default for Machine<V> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<V: AbstractValue> Machine<V> {
+    /// Provide initial values for some wires.
+    // Likely for test purposes only - should we make non-pub or #[cfg(test)] ?
```

Comment:
Not sure quite what to do here. @doug-q may have opinions?

Reply:
We think this potentially useful (albeit not very powerful, as any prepopulated value will be `join`ed with anything computed). A version that "breaks" the wire (replacing any value computed, without `join`) would be more effective, also dangerous....